### DISCUSSION TITLE
file outline slow

### DISCUSSION THREAD

**Discussion by amigo421** [microsoft/vscode-cpptools#7570] [Q&A]
I've opened the project as folder , all required paths added to includePath.

when I open some file ~500 lines, Loading document symbols for ... takes an hours!!

"Tag parser initializing.." also for an hours and nothing more happens.
so no any symbol can be resolved

last version of VS from stable branch (auto-updated)
remote linux mode WIn10=>RHEL7

also  I see this in the log: IntelliSense client not available, using Tag Parser for go to definition

**Comment by sean-mcmanus** [comment]
If you don't open a file and instead run C/C++: Rescan Workspace does it still take a long time for the database status icon to disappear? Or if you open a file with just "int main() { return 0; }"?

Are you using WSL1 or WSL2 and are you files resident on in the WSL VM or on the Windows /mnt/?

**Reply by amigo421** [reply]
Thank you for assistance !
I'm using RedHat on real remote host , not any WSL
I've started rescan ,
two icaons:
burning - "updating intellisense "
and storage - "parsing paused" (why paused??)


it takes a few hours , I'll update about the results 

opening trivial projects doesn't have such issues

**Reply by amigo421** [reply]
> If you don't open a file and instead run C/C++: Rescan Workspace does it still take a long time for the database status 

in addition to my previous reply: I've click : resume parsing and after a few hours database icon is still there and tip is: Parsing Files : 0 / 23016 (0%). For sure - for now, status is a parsing, not paused
I'll keep this overnight , but you see - the work is impossible with such speed

**Reply by sean-mcmanus** [reply]
What do you see in the C/C++ logging when you set the C_Cpp.loggingLevel to "Debug"? Do you see "Tag parsing" messages? Can you temporarily set your compilerPath to "" to see if that helps?

**Reply by amigo421** [reply]
it was already with with debug level, 

it hangs somewhere in boost, last message is a:

> tag parsing file: ...../include/boost/typeof/vector200.hpp

and yes, I'll clean this variable and in case of success I'll add a paths one-by-one to catch the problem

**Reply by sean-mcmanus** [reply]
You could also try adding folders to your C_Cpp.files.exclude setting to stop parsing as many files, e.g.
```
    "C_Cpp.files.exclude": {
        "/usr/include/boost": true,
        "**/.vscode": true
    }
```
depending on what logging you are seeing.

It's possible a symlink to a very large folder could exist, so adding that symlink to the C_Cpp.files.exclude could also fix that.

**Reply by amigo421** [reply]
thank you , I'll try this technique as well 

but for now , could you clarify me please,
when I call workspace rescan , all folders included into this will be rescaned, right? 

probably you remember , I was asking you some time ago about how to add search the files and symbols in Ctrl+P/O , and the adding all "includes" into the workspace was recommended . 

so now , some include folders exist as "includePath" as well as added into the workspace for the navigation.

looks like double parsing? probably this is a root of the issue.

by the way, SSH FS extension adds remote file systems into the my workspace, is it also in parsing procedure?

**Reply by sean-mcmanus** [reply]
It should rescan all the folders included in the browse.path setting (which by default includes folders from includePath and from querying the compilerPath as well) but excluding files.exclude paths.

We have logic that should avoid double scanning. Your logging should show messages like "Processing folder" that shows what folders are being processed.

All processing is done on the remote machine.

**Reply by amigo421** [reply]
could you explain please what does "Configuration for file ... was not found" mean? 
a  number of such messages in C++ log...

**Reply by amigo421** [reply]
after the removing all external paths it parses the files but still veeeery slowly some files requires a min, some ones -10 mins,
but the problem is not this for the moment.
unfortunately it parses wrongly. 
even basic parse by classes and methods is incorrect. "outline" of the file doesn't reflect real structure.
e.g. it doesn't see one method in the class but "think" that macro is a method (really - macro is inside the method and expands into range loop)

**Reply by sean-mcmanus** [reply]
"Configuration for file...was not found" can occur if you have compileCommands or configurationProvider set but open a file that doesn't have configuration info from those sources.

I'm not sure what could be causing the slow parsing...is the remote machine slow? Are the slow parses on small or large files?

The incorrect parsing sounds like it could be caused by defines that break our parser. Any defines that need to evaluated in order to parse correctly need to be added to a cpp.hint file (https://docs.microsoft.com/en-us/cpp/build/reference/hint-files?view=msvc-160).
